src/drivers/amd/rembrandt/acp_sw_audio_dma.c

```diff
@@ -0,0 +1,640 @@
+	key = k_spin_lock(&channel->dma->lock);
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+	k_spin_unlock(&channel->dma->lock, key);
+}
+
+static int acp_dai_sw_audio_dma_start(struct dma_chan_data *channel)
+{
+	uint32_t		sw0_audio_tx_en = 0;
+	uint32_t		sw0_audio_rx_en = 0;
+	uint32_t		acp_pdm_en;
+	int    tx_en_status;
+	int    rx_en_status;
+	uint32_t delay_cnt = 10000;
+	int i = 0;
+
+	sw0_audio_tx_en = 0;
+	sw0_audio_rx_en = 0;
+	for (i = 0; i < 8; i = i + 2) {
+		sw0_audio_tx_en |= io_reg_read((PU_REGISTER_BASE + sw_dev[i].sw_dev_en));
+		sw0_audio_rx_en |= io_reg_read((PU_REGISTER_BASE + sw_dev[i + 1].sw_dev_en));
+	}
+
+	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
+	if (!sw0_audio_tx_en && !sw0_audio_rx_en && !acp_pdm_en) {
+		/* Request SMU to set aclk to 600 Mhz */
+		acp_change_clock_notify(600000000);
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), ACP_ACLK_CLK_SEL);
+	}
+
+	if (channel->direction == DMA_DIR_MEM_TO_DEV) {
+		channel->status = COMP_STATE_ACTIVE;
+
+		io_reg_write((PU_REGISTER_BASE + sw_dev[channel->index].sw_dev_en), 1);
+		tx_en_status = io_reg_read(PU_REGISTER_BASE +
+		sw_dev[channel->index].sw_dev_en_status);
+		while (delay_cnt > 0) {
+			tx_en_status = io_reg_read(PU_REGISTER_BASE +
+			sw_dev[channel->index].sw_dev_en_status);
+			if ((tx_en_status & 0x1) == 1)
+				break;
+			delay_cnt--;
+		}
+	} else if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		channel->status = COMP_STATE_ACTIVE;
+
+		io_reg_write((PU_REGISTER_BASE + sw_dev[channel->index].sw_dev_en), 1);
+
+		rx_en_status = io_reg_read(PU_REGISTER_BASE +
+		sw_dev[channel->index].sw_dev_en_status);
+		while (delay_cnt > 0) {
+			rx_en_status = io_reg_read(PU_REGISTER_BASE +
+			sw_dev[channel->index].sw_dev_en_status);
+			if ((rx_en_status & 0x1) == 1)
+				break;
+			delay_cnt--;
+		}
+	} else {
+		tr_err(&acp_sw_audio_tr, "Start direction not defined %d", channel->direction);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int acp_dai_sw_audio_dma_release(struct dma_chan_data *channel)
+{
+	/* nothing to do on rembrandt */
+	return 0;
+}
+
+static int acp_dai_sw_audio_dma_pause(struct dma_chan_data *channel)
+{
+	/* nothing to do on rembrandt */
+	return 0;
+}
+
+static int acp_dai_sw_audio_dma_stop(struct dma_chan_data *channel)
+{
+	uint32_t		sw0_audio_tx_en = 0;
+	uint32_t		sw0_audio_rx_en = 0;
+	uint32_t		acp_pdm_en;
+	int    tx_en_status;
+	int    rx_en_status;
+	uint32_t delay_cnt = 10000;
+	int i;
+
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 0;
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	if (channel->direction == DMA_DIR_MEM_TO_DEV) {
+		io_reg_write((PU_REGISTER_BASE + sw_dev[channel->index].sw_dev_en), 0);
+		tx_en_status = io_reg_read(PU_REGISTER_BASE +
+		sw_dev[channel->index].sw_dev_en_status);
+		while (delay_cnt > 0) {
```

Comment:
similar comments about `switch`, indentation, status reading and loops as above

Reply:
done